nekoyume/Assets/_Scripts/BlockChain/ActionRenderHandler.cs

```diff
@@ -452,199 +467,142 @@ private void ResponseSell(ActionBase.ActionEvaluation<Sell> eval)
 
-                            var purchaseHistory = ReactiveShopState.PurchaseHistory[eval.Action.Id];
-                            var item = purchaseHistory.FirstOrDefault(x => x.ProductId.Value == purchaseResult.productId);
-                            if (item is null)
-                            {
-                                continue;
-                            }
+            var agentAddress = States.Instance.AgentState.address;
+            var avatarAddress = States.Instance.CurrentAvatarState.address;
+            if (!eval.OutputStates.TryGetAvatarStateV2(agentAddress, avatarAddress,
+                out var avatarState))
+            {
+                return;
+            }
 
-                            // Local layer
-                            var price = item.Price.Value;
-                            LocalLayerModifier.ModifyAgentGold(agentAddress, price);
-
-                            // Push notification
-                            var errorType = ((ShopErrorType) purchaseResult.errorCode).ToString();
-                            var msg = string.Format(L10nManager.Localize("NOTIFICATION_BUY_FAIL"),
-                                item.ItemBase.Value.GetLocalizedName(),
-                                L10nManager.Localize(errorType),
-                                price);
-                            OneLinePopup.Push(MailType.Auction, msg);
-                        }
-                    }
+            var errors = eval.Action.errors.ToList();
+            var purchaseInfos = eval.Action.purchaseInfos;
+            if (eval.Action.buyerAvatarAddress == avatarAddress) // check error
+            {
+                foreach (var (orderId, errorCode) in errors)
+                {
+                    var order = Util.GetOrder(orderId);
+                    var itembase = Util.GetItemBaseByOrderId(orderId);
+                    var price = order.Price;
+                    var errorType = ((ShopErrorType) errorCode).ToString();
+                    LocalLayerModifier.ModifyAgentGold(agentAddress, price);
+                    var msg = string.Format(L10nManager.Localize("NOTIFICATION_BUY_FAIL"),
+                                                  itembase.GetLocalizedName(),
+                                                  L10nManager.Localize(errorType),
+                                                  price);
+                    OneLinePopup.Push(MailType.Auction, msg);
                 }
-                else
+            }
+
+            foreach (var purchaseInfo in purchaseInfos)
+            {
+                if (!errors.Exists(tuple => tuple.orderId.Equals(purchaseInfo.OrderId)))
                 {
-                    var buyerAvatarAddress = eval.Action.buyerAvatarAddress;
-                    var buyerAvatarStateValue = eval.OutputStates.GetState(buyerAvatarAddress);
-                    if (buyerAvatarStateValue is null)
+                    if (eval.Action.buyerAvatarAddress == avatarAddress) // buyer
```

Comment:
This `if` statement is better wrapped around the outer `foreach` statement like below.
```
if (eval.Action.buyerAvatarAddress == avatarAddress) // buyer
{
  foreach (var purchaseInfo in purchaseInfos)
  {
    if (!errors.Exists(tuple => tuple.orderId.Equals(purchaseInfo.OrderId)))
    {
    }
  }
}
else // seller
{
  foreach (var purchaseInfo in purchaseInfos)
  {
    if (!errors.Exists(tuple => tuple.orderId.Equals(purchaseInfo.OrderId)))
    {
    }
  }
}
```

Reply:
fixed👍🏻